### PR TITLE
disttask: add notify when task submitted or changed & refactor test

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:

### What changed and how does it work?
- merge & refactor framework test
- notify task change when new task is submitted or finishes, and wait task asyncly. now tests in `framework_test.go` can run in 1:30, it takes 2:30 previously.
- rename ExecuteTask in scheduler to ScheduleTask
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/index.go [hunk 1]
```diff
@@ -2209,6 +2209,41 @@ func (w *worker) updateJobRowCount(taskKey string, jobID int64) {
 	w.getReorgCtx(jobID).setRowCount(rowCount)
 }
 
+// submitAndWaitTask submits a task and wait for it to finish.
+func submitAndWaitTask(ctx context.Context, taskKey string, taskType proto.TaskType, concurrency int, taskMeta []byte) error {
```
Comment (D3Hunter) [inline pkg/ddl/index.go:2213]: moved from `handle.go`

File: pkg/disttask/framework/framework_err_handling_test.go [hunk 2]
```diff
@@ -19,31 +19,23 @@ import (
...
 	ctx, ctrl, testContext, distContext := testutil.InitTestContext(t, 2)
 	defer ctrl.Finish()
 	testutil.RegisterTaskMeta(t, ctrl, testutil.GetPlanErrSchedulerExt(ctrl, testContext), testContext, nil)
-	testutil.DispatchTaskAndCheckSuccess(ctx, t, "key1", testContext, nil)
-	testContext.CallTime = 0
+	submitTaskAndCheckSuccessForBasic(ctx, t, "key1", testContext)
 	distContext.Close()
 }
 
-func TestRevertPlanErr(t *testing.T) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/framework_err_handling_test.go:33]: this test duplicate with previous one

File: pkg/disttask/framework/framework_test.go [hunk 3]
```diff
@@ -24,64 +27,90 @@ import (
...
-	testutil.DispatchTaskAndCheckSuccess(ctx, t, "key5", testContext, nil)
-	distContext.Close()
+func submitTaskAndCheckSuccessForBasic(ctx context.Context, t *testing.T, taskKey string, testContext *testutil.TestContext) {
+	submitTaskAndCheckSuccess(ctx, t, taskKey, testContext, map[proto.Step]int{
+		proto.StepOne: 3,
+		proto.StepTwo: 1,
+	})
 }
 
-func TestFramework3Server(t *testing.T) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/framework_test.go:47]: very similar with previous one, so merge them into `TestRandomOwnerChangeWithMultipleTasks`

File: pkg/disttask/framework/framework_test.go [hunk 4]
```diff
@@ -213,17 +261,6 @@ func TestFrameworkSetLabel(t *testing.T) {
 	distContext.Close()
 }
 
-func TestMultiTasks(t *testing.T) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/framework_test.go:216]: similar with `TestRandomOwnerChangeWithMultipleTasks`, so remove this one

File: pkg/disttask/framework/testutil/disttest_util.go [hunk 5]
```diff
@@ -115,105 +110,22 @@ func RegisterRollbackTaskMeta(t *testing.T, ctrl *gomock.Controller, mockSchedul
...
-		require.NotNil(t, task)
-		if task.State != proto.TaskStatePending && task.State != proto.TaskStateRunning && task.State != proto.TaskStateCancelling && task.State != proto.TaskStateReverting && task.State != proto.TaskStatePausing {
-			break
-		}
-	}
-	return task
-}
-
-// DispatchTaskAndCheckSuccess schedule one task and check if it is succeed.
-func DispatchTaskAndCheckSuccess(ctx context.Context, t *testing.T, taskKey string, testContext *TestContext, checkResultFn func(t *testing.T, testContext *TestContext)) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/testutil/disttest_util.go:150]: inlined, or removed

File: pkg/disttask/framework/scheduler/scheduler.go [hunk 6]
```diff
@@ -528,7 +528,7 @@ func (s *BaseScheduler) onErrHandlingStage(receiveErrs []error) error {
 
 func (s *BaseScheduler) onNextStage() (err error) {
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/scheduler.go:529]:
```suggestion
func (s *BaseScheduler) onNextStep() (err error) {
```

**Review by ywqzzy — APPROVED** [review]

File: pkg/disttask/framework/scheduler/scheduler.go [hunk 7]
```diff
@@ -541,16 +543,6 @@ func (s *BaseScheduler) onNextStage() (err error) {
 		return s.taskMgr.SucceedTask(s.ctx, s.Task.ID)
 	}
 
-	// Adjust the task's concurrency.
-	if s.Task.State == proto.TaskStatePending {
```
Comment (okJiang) [inline pkg/disttask/framework/scheduler/scheduler.go:545]: why delete these?

**Review by okJiang — APPROVED** [review]
The changes made in this PR are too much for me, maybe they could be split up next time.😵‍💫

File: pkg/disttask/framework/scheduler/scheduler.go (hunk 7 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/scheduler.go:545]: concurrency shouldn't adjust in scheduler

**Comment by D3Hunter** [comment]
> The changes made in this PR are too much for me, maybe they could be split up next time.😵‍💫

without refactor some test, we cannot add notify easily. most code change in this pr are tests

**Review by ywqzzy — APPROVED** [review]

**Pull request merged**